### PR TITLE
Implement the IgnoreReadOnlyProperties option

### PR DESCRIPTION
**Pull request by justing-bq**
Resolves https://github.com/amzn/ion-object-mapper-dotnet/issues/7.

Option to determine whether or not to ignore read only properties during serialization/deserialization.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 1]
```diff
@@ -173,6 +194,12 @@ private bool IsField(FieldInfo field)
                 return true;
             }
 
+            if (!options.IgnoreReadOnlyProperties && 
+                readOnlyProperties.Any(p => field.Name == $"<{p.Name}>k__BackingField"))
```
Comment (simonz-bq) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:198]: This seems finnicky. I'm not 100% sure what it's for but is there a potentially a better way to do this?

**Review by simonz-bq — CHANGES_REQUESTED** [review]
@justing-bq please add stylecop to the project asap, and configure your IDE to do some basics such as line length checking and blank space removal on save.

Inconsistencies such as a lack of "this" or missing line breaks after {} blocks are starting to appear.

File: Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs [hunk 2]
```diff
@@ -24,13 +25,14 @@ public void SerializesAndDeserializesObjectsWithIncludeFields()
         public void SerializesObjectsWithIgnoreNulls()
         {
             var serializer = new IonSerializer(new IonSerializationOptions {IgnoreNulls = true});
-            var motorcycle = new Motorcycle {canOffroad = true};
+            var motorcycle = new Motorcycle(null, null, true, DateTime.Now);
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs:28]: Why this change? I believe it's because some properties no longer have setters?

File: Amazon.Ion.ObjectMapper.Test/TestObjects.cs [hunk 3]
```diff
@@ -114,6 +132,8 @@ public static class TestObjects {
             Weight = new Random().NextDouble(),
             Engine = new Engine { Cylinders = 4, ManufactureDate = DateTime.Parse("2009-10-10T13:15:21Z") }
         };
+        
+        public static Motorcycle harley = new Motorcycle("Harley", "Black", true, DateTime.MinValue);
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper.Test/TestObjects.cs:136]: Should be readonly and variable name should be all caps - unless for some reason it's using a different convention here?

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 4]
```diff
@@ -12,12 +12,18 @@ public class IonObjectSerializer : IonSerializer<object>
...
         private readonly Type targetType;
+        private readonly IEnumerable<PropertyInfo> readOnlyProperties;
 
         public IonObjectSerializer(IonSerializer ionSerializer, IonSerializationOptions options, Type targetType)
         {
             this.ionSerializer = ionSerializer;
             this.options = options;
             this.targetType = targetType;
+
+            if (!this.options.IgnoreReadOnlyProperties && !this.options.IncludeFields)
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:23]: Shouldn't this read the properties only if you are including fields?

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 5]
```diff
@@ -32,8 +38,12 @@ public object Deserialize(IIonReader reader)
                 FieldInfo field;
                 if (property != null)
                 {
+                    if (!property.CanWrite)
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:41]: So read-only properties are never set on deserialization? Even if they were serialized?

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 6]
```diff
@@ -88,12 +102,13 @@ public void Serialize(IIonWriter writer, object item)
 
             foreach (var field in Fields())
             {
-                var fieldValue = field.GetValue(item);
-                if (options.IgnoreNulls && fieldValue == null)
+                if (options.IgnoreReadOnlyFields && field.IsInitOnly)
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:105]: Why the order change of the if statements here?

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 1 above)
Comment (kylepbit) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:198]: I agree - this seems not great, are there alternatives?

File: Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs (hunk 2 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs:28]: Constructors are required to set the readonly property `PurchaseDate`.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 5 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:46]: The issue is that `property.SetValue()` will not work for a read-only property because it has no setter.
To get around this we instead deserialize the read-only property's backing field. So the logic for handling read-only properties strangely happens during field processing instead of property processing.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 4 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:23]: So if we are serializing/deserializing read-only properties then we need to serialize/deserialize the property's backing field (and that is what `this.readOnlyProperties` is used for).
However the `IncludeFields` option results in all fields (including backing fields) being included which eliminates any need for special logic to detect those backing fields.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 1 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs]: I wish but I couldn't figure out another way to detect that a given field is a backing field for a given property other than this name comparison.

**Review by HomayounA — COMMENTED** [review]
No other comments.

File: Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs (hunk 2 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs]: Upon further investigation, the `{ field = value }` style is allowed if there is a parameterless constructor. It was only a problem when I first introduced the parameterized constructor before adding the parameterless constructor.
I have now reverted the style back to what it was before.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 7]
```diff
@@ -137,6 +160,12 @@ private PropertyInfo FindProperty(string readName)
             var name = options.NamingConvention.ToProperty(readName);
             return targetType.GetProperty(name);
         }
+
+        private IEnumerable<PropertyInfo> ReadOnlyProperties(Type targetType)
+        {
+            return targetType.GetProperties().Where(p => !p.CanWrite);
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:166]: This does not define a readonly property, this is simply a property with no setter. See this link for a description of how to find readonly properties: https://stackoverflow.com/questions/5243764/at-runtime-how-can-i-test-whether-a-property-is-readonly.

**Review by alpian — CHANGES_REQUESTED** [review]

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 7 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs]: Addressed.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 8]
```diff
@@ -137,6 +161,13 @@ private PropertyInfo FindProperty(string readName)
             var name = options.NamingConvention.ToProperty(readName);
             return targetType.GetProperty(name);
         }
+
+        private bool IsReadOnlyProperty(PropertyInfo prop)
+        {
+            return !prop.CanWrite ||
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:173]: I think this is still incorrect. I think you should just remove the `!prop.CanWrite` part.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 9]
```diff
@@ -12,12 +13,23 @@ public class IonObjectSerializer : IonSerializer<object>
...
             this.options = options;
             this.targetType = targetType;
+            
+            if (!this.options.IgnoreReadOnlyProperties && !this.options.IncludeFields)
+            {
+                // If we are serialzing/deserializing readonly properties then we need to serialize/deserialize those
+                // properties' backing fields.
+                // this.readOnlyProperties is used for detecting the appropriate backing fields.
+                // However if we are already including all fields,
+                // then there is no need for this backing field detection logic.
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:30]: I don't think the constructor should have this type of logic in it. Rather do this in the Serialize

**Review by alpian — CHANGES_REQUESTED** [review]

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 9 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs]: Now this item is lazily instantiated so it only creates if/when we need it.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 8 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs]: After removing `!prop.CanWrite`, a property with a `{ get; }` accessor does not pass `IsReadOnlyProperty()` because it fails the `ReadOnlyAttribute` check.

So now I've redefined this to `property.SetMethod == null`.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 10]
```diff
@@ -32,8 +35,15 @@ public object Deserialize(IIonReader reader)
                 FieldInfo field;
                 if (property != null)
                 {
+                    if (IsReadOnlyProperty(property))
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:38]: `IgnoreReadOnlyProperties` only applies to Serialization.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 11]
```diff
@@ -179,6 +200,12 @@ private bool IsField(FieldInfo field)
                 return true;
             }
 
+            if (!this.options.IgnoreReadOnlyProperties &&
```
Comment (alpian) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:203]: How is this related to this PR? Seems like maybe this needs to be rebased?

**Review by alpian — CHANGES_REQUESTED** [review]

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 10 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:38]: We still need this check regardless of the `IgnoreReadOnlyProperties` option because without it `property.SetValue()` will throw an exception on the read-only property.

Because of this check we still need to handle the scenario where we are deserializing read-only properties (when the `IgnoreReadOnlyProperties` option is false). Instead of calling `property.SetValue()`, we deserialize the corresponding backing field instead.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 11 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:203]: This logic is for detecting backing fields for readonly properties since we can't deserialize read-only properties the normal way. 
It's still related to this PR because this PR introduced the `if (IsReadOnlyProperty(property))` check in `Deserialize()`
